### PR TITLE
Add compel node and conditioning field type

### PR DESCRIPTION
**Pull request by StAlKeR7779**
Done as I said in title, but need to test(and understand) how cli works, as previously it uses single prompt and now it's positive and negative.

### CODE REVIEW

**Comment by blessedcoolant** [comment]
Works as advertised.

But I think we need to change a few things unrelated to this PR.

1. Get a model loader node in which outputs three values - `Text Encoder`, `UNet`, `VAE`
2. Feed the `Text Encoder` to the Compel Node instead of the `Model`
3. Feed the `UNet` to the Text To Latent instead of `Model`
4. Feed the `VAE` to the Latent instead of the `Model`

**Review by damian0815 — CHANGES_REQUESTED** [review]

**Review by damian0815 — CHANGES_REQUESTED** [review]

File: invokeai/app/invocations/compel.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        model = choose_model(context.services.model_manager, self.model)
+        pipeline = model["model"]
+        tokenizer = pipeline.tokenizer
+        text_encoder = pipeline.text_encoder
+
+        # TODO: global? input?
+        #use_full_precision = precision == "float32" or precision == "autocast"
+        use_full_precision = False
+
+        textual_inversion_manager = TextualInversionManager(
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/app/invocations/compel.py]: Yeah, it's copypaste from my other draft)
Changed to textural inversion model from loaded model for now.
But we need to rewrite it when partial model loading implemented by lstein.

File: invokeai/app/invocations/compel.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+        if not self.truncate_long_prompts:
+            [c, uc] = compel.pad_conditioning_tensors_to_same_length([c, uc])
+
+        c_ec = InvokeAIDiffuserComponent.ExtraConditioningInfo(
+            tokens_count_including_eos_bos=get_max_token_count(tokenizer, positive_prompt),
+            cross_attention_control_args=c_options.get("cross_attention_control", None),
+        )
+
+        uc_ec = InvokeAIDiffuserComponent.ExtraConditioningInfo(
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/app/invocations/compel.py]: Done

**Review by damian0815 — APPROVED** [review]

File: invokeai/app/invocations/compel.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+class CompelOutput(BaseInvocationOutput):
+    """Compel parser output"""
+
+    #fmt: off
+    type: Literal["compel_output"] = "compel_output"
+    # name + loras -> pipeline + loras
+    # model: ModelField           = Field(default=None, description="Model")
+    # src? + loras -> tokenizer + text_encoder + loras
+    # clip:  ClipField            = Field(default=None, description="Text encoder(clip)")
+    positive: ConditioningField = Field(default=None, description="Positive conditioning")
```
Comment (Kyle0654) [inline invokeai/app/invocations/compel.py:37]: Is there a use case for separating these? Could you generate just one of them and then re-use that? Would it make sense to have separate nodes, or maybe a positive/negative parameter on the node to generate these separately? Do these work correctly later on if used with a different model?

File: invokeai/app/invocations/compel.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    negative: ConditioningField = Field(default=None, description="Negative conditioning")
+    #fmt: on
+
+
+class CompelInvocation(BaseInvocation):
+
+    type: Literal["compel"] = "compel"
+
+    positive_prompt: str = Field(default="", description="Positive prompt")
```
Comment (Kyle0654) [inline invokeai/app/invocations/compel.py:46]: Kind of same comment as above - it'd potentially be nice if this just generated a single set of conditioning (positive or negative). Then the parameter could just be "prompt".

Also, does compel not process negative prompts using brackets anymore? I haven't been using it much in a while - has our prompt syntax changed?

File: invokeai/app/invocations/latent.py [hunk 5]
```diff
@@ -143,9 +143,9 @@ class TextToLatentsInvocation(BaseInvocation):
     type: Literal["t2l"] = "t2l"
 
     # Inputs
-    # TODO: consider making prompt optional to enable providing prompt through a link
     # fmt: off
-    prompt: Optional[str] = Field(description="The prompt to generate an image from")
+    positive: Optional[ConditioningField] = Field(description="Positive conditioning for generation")
```
Comment (Kyle0654) [inline invokeai/app/invocations/latent.py:147]: Would `conditioning` and `unconditioning` be more accurate field names?

Comment (Kyle0654) [reply invokeai/app/invocations/latent.py:147]: I also think you may need to add some schema config for the UI so it knows what's actually required as input here (e.g. positive conditioning may be required, whereas negative conditioning may actually be optional?)

**Review by Kyle0654 — COMMENTED** [review]
Some questions and comments about usability. Try using it in the CLI to validate usability =)

File: invokeai/app/invocations/latent.py (hunk 5 above)
Comment (psychedelicious) [reply invokeai/app/invocations/latent.py:147]: (I'm going to refactor the schema config structure now that we've played around with this a bit and the needs are clearer. When I do that, I'll update all nodes' Config to match.)

Comment (StAlKeR7779) [reply invokeai/app/invocations/latent.py:147]: Ok will add schema, about rename - not sure. From code perspective it's more accurate, but users see always "positive/negative prompt", so they don't know conditional/unconditional terms)

File: invokeai/app/invocations/compel.py (hunk 4 above)
Comment (StAlKeR7779) [reply invokeai/app/invocations/compel.py:46]: Compel just translates prompt string to conditioning tensor, it's don't know is this tensor from positive prompt or from negative. It's invoke parses [] now.
So, if do more precisely we need to do compel with only 1 field and user need to create 2 compel nodes: first - for positive prompt parsing and second for negative prompt parsing.
I also firstly thinks about only 1 prompt input and output(so user will create 2 compel nodes), but now we not sure how loras will work and if loras will be applied in compel from prompts it's better to be 2 inputs in node as we can easily parse them with same text_encoder state(and that also why in my comments compel has output unet and clip(text_encoder)).

File: invokeai/app/invocations/latent.py (hunk 5 above)
Comment (psychedelicious) [reply invokeai/app/invocations/latent.py:147]: +1 to teaching the users the correct terminology (eg conditioning and unconditioning)

Comment (damian0815) [reply invokeai/app/invocations/latent.py:147]: ehhh. if you put a prompt in the “unconditioned conditioning” tensor it’s no longer “unconditioned”. i think positive and negative are better terms at this point

Comment (StAlKeR7779) [reply invokeai/app/invocations/latent.py]: In terms of perp-neg, for example, unconditioning tensor it's empty tensor, and for prompts it's positive and negative tensors. Also I combined it in one field in last commit as without it it's hard to imagine how to implement perp-neg option.

File: invokeai/app/invocations/latent.py [hunk 6]
```diff
@@ -143,9 +143,9 @@ class TextToLatentsInvocation(BaseInvocation):
     type: Literal["t2l"] = "t2l"
 
     # Inputs
-    # TODO: consider making prompt optional to enable providing prompt through a link
     # fmt: off
-    prompt: Optional[str] = Field(description="The prompt to generate an image from")
+    conditioning: Optional[ConditioningField] = Field(description="Conditioning for generation")
+    seed:        int = Field(default=-1,ge=-1, le=np.iinfo(np.uint32).max, description="The seed to use (-1 for a random seed)", )
```
Comment (Kyle0654) [inline invokeai/app/invocations/latent.py:148]: We've removed seed from this node (since it's only used in the noise node).

**Review by Kyle0654 — CHANGES_REQUESTED** [review]

File: invokeai/app/invocations/latent.py (hunk 6 above)
Comment (StAlKeR7779) [reply invokeai/app/invocations/latent.py:148]: Ah... A bit wrong merge conflict resolve. I'l delete it.

**Review by Kyle0654 — APPROVED** [review]

**Comment by blessedcoolant** [comment]
Functionality is good. But why did you merge both Positive and Negative conditioning into the same output? I think they should both be outputted as individual outputs so they can be plugged as needed across multiple nodes if need be.

File: invokeai/app/invocations/compel.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    # name + loras -> pipeline + loras
+    # model: ModelField = Field(default=None, description="Model to use") 
+    # src? + loras -> tokenizer + text_encoder + loras
+    # clip: ClipField = Field(default=None, description="Text encoder(clip) to use")
+
+    # Schema customisation
+    class Config(InvocationConfig):
+        schema_extra = {
+            "ui": {
+                "tags": ["latents", "noise"],
```
Comment (blessedcoolant) [inline invokeai/app/invocations/compel.py:61]: These tags need to be updated to be more appropriate.

File: invokeai/app/invocations/compel.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    # name + loras -> pipeline + loras
+    # model: ModelField           = Field(default=None, description="Model")
+    # src? + loras -> tokenizer + text_encoder + loras
+    # clip:  ClipField            = Field(default=None, description="Text encoder(clip)")
+    positive: ConditioningField = Field(default=None, description="Positive conditioning")
+    negative: ConditioningField = Field(default=None, description="Negative conditioning")
+    #fmt: on
+
+
+class CompelInvocation(BaseInvocation):
```
Comment (blessedcoolant) [inline invokeai/app/invocations/compel.py:42]: This class needs a description so it is available in the info tag in the UI.

File: invokeai/app/invocations/compel.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+            cross_attention_control_args=options.get("cross_attention_control", None),
+        )
+
+        name_prefix = f'{context.graph_execution_state_id}__{self.id}'
+        name_positive = f"{name_prefix}_positive"
+        name_negative = f"{name_prefix}_negative"
+
+        # TODO: hacky but works ;D maybe rename latents somehow?
+        context.services.latents.set(name_positive, (c, ec))
+        context.services.latents.set(name_negative, (uc, None))
```
Comment (blessedcoolant) [inline invokeai/app/invocations/compel.py:143]: Is this the best way to forward this data? Because every time a compel node is run, the latent storage will save a file to the disk. Not sure if this should be the intended behavior. But this is something @Kyle0654 should be the judge of.

**Review by blessedcoolant — CHANGES_REQUESTED** [review]

File: invokeai/app/invocations/compel.py (hunk 7 above)
Comment (blessedcoolant) [inline invokeai/app/invocations/compel.py:61]: As discussed on Discord, think we should it to Prompt (Compel) with `"title"` in the tag.

**Review by blessedcoolant — CHANGES_REQUESTED** [review]

File: invokeai/app/invocations/compel.py (hunk 9 above)
Comment (StAlKeR7779) [reply invokeai/app/invocations/compel.py]: As mentioned in discord - it's implies by architecture to save all results, so next node can run in other process/machine.

**Review by blessedcoolant — APPROVED** [review]
